From: Michael Weghorn <m.weghorn@posteo.de>
Date: Tue, 30 Sep 2025 23:10:46 +0200
Subject: a11y: Don't send focus-related events for unfocused treeview

When the GtkTreeView doesn't have focus, skip
sending a11y events that are only applicable
if it had focus.

This code is triggered e.g. when calling
gtk_tree_view_set_cursor, and sending the events
for an unfocused tree view could break focus
tracking done by assistive technologies, as
the events would suggest that one of the treeview's
cells actually has keyboard focus when it doesn't.

Bug: https://gitlab.gnome.org/GNOME/gtk/-/issues/7796
Origin: upstream, 3.24.52, commit:fbc28541f0e33754ee88f6241604a06594349fa9
---
 gtk/a11y/gtktreeviewaccessible.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/gtk/a11y/gtktreeviewaccessible.c b/gtk/a11y/gtktreeviewaccessible.c
index c2b7e8a..7341e72 100644
--- a/gtk/a11y/gtktreeviewaccessible.c
+++ b/gtk/a11y/gtktreeviewaccessible.c
@@ -2028,6 +2028,9 @@ _gtk_tree_view_accessible_add_state (GtkTreeView          *treeview,
 
   if (state == GTK_CELL_RENDERER_FOCUSED)
     {
+      if (!gtk_widget_has_focus (GTK_WIDGET (treeview)))
+        return;
+
       single_column = get_effective_focus_column (treeview, _gtk_tree_view_get_focus_column (treeview));
     }
   else if (state == GTK_CELL_RENDERER_EXPANDED ||
